### ISSUE TITLE
Pub/Sub ISubscriber(s) callbacks do not triggered anymore after ISubscriber unsubscribes

### ISSUE THREAD

**Issue by sanzor** [StackExchange/StackExchange.Redis#1648]
I have a situation where i use multiple `ISubscribers` to a Redis channel and when one of them uses ISubscribers.UnsubscribeAsync/Unsubscribe/UnsubscribeAll/UnsubscribeAllAsync all other's callbacks do not trigger anymore.

I have tried the below scenario:

- ISubscriber 1 subscribes to channel  
- ISubscriber 2 subscribes to channel
- Subscribe from `Redis-Cli` as an independent observer to target channel
- Start publishing messages to channel (via a redis-cli)
- Call ISubscriber 1 Unsubscribe (any of the 4 variants will work)
-  All other ISubscribers to target channel do not get their callback triggered anymore.

Below is the code to test:  (You also need some way to publish messages - i used a redis-cli)
```
class Program {
        public static string CHANNEL = "chan";
        public static TaskCompletionSource<bool> tcs = new System.Threading.Tasks.TaskCompletionSource<bool>();
        public static TaskCompletionSource<bool> tcs2 = new TaskCompletionSource<bool>();
        public static int INITIAL_COUNT = 3;
        static async Task Main(string[] args) {
            List<ISubscriber> subs = new List<ISubscriber>();
            
            ConnectionMultiplexer mux = ConnectionMultiplexer.Connect("localhost:6379");
            for (int i = 0; i < INITIAL_COUNT; i++) {
                var sub = mux.GetSubscriber();
                int closuredIdentity = i;
                await sub.SubscribeAsync(CHANNEL, async (chan, msg) => {
                   if(msg=="k" && closuredIdentity == 0) {
                        Console.WriteLine($"Subscriber[{closuredIdentity}] unsubscribes");
                        await sub.UnsubscribeAsync(CHANNEL);
                        tcs.SetResult(true);
                    }
                    Console.WriteLine($"From sub[{closuredIdentity}]:{msg}");
                });
                subs.Add(sub);
            }
            var destroyerSubscriber = subs.FirstOrDefault();
            await tcs.Task;
            Console.WriteLine("Messages post unsubscribe");

            await destroyerSubscriber.SubscribeAsync(CHANNEL, async (chan, msg) => {
                if (msg == "k") {
                    tcs2.SetResult(true);
                }
            });
            await tcs2.Task;
            Console.WriteLine("Hello World!");
        }

    }


```

**Comment by mgravell** [comment]
To quote from the intellisense:

         /// <summary>
        /// Unsubscribe from a specified message channel; note; if no handler is specified, the subscription is cancelled regardless
        /// of the subscribers; if a handler is specified, the subscription is only cancelled if this handler is the 
        /// last handler remaining against the channel
        /// </summary>
        /// <param name="channel">The channel that was subscribed to.</param>
        /// <param name="handler">The handler to no longer invoke when a message is received on <paramref name="channel"/>.</param>
        /// <param name="flags">The command flags to use.</param>
        /// <remarks>https://redis.io/commands/unsubscribe</remarks>
        /// <remarks>https://redis.io/commands/punsubscribe</remarks>
        void Unsubscribe(RedisChannel channel, Action<RedisChannel, RedisValue> handler = null, CommandFlags flags = CommandFlags.None);

Basically, to get what you want (tracked unsubscription), you should supply the same handler to `Unsubscribe` as you did to `Subscribe`; when the system determines that all handles have been removed: it will nuke the subscription. By calling `Unsubscribe[Async]` without a handler, you're issuing a blanket unsubscribe.

**Comment by sanzor** [comment]
Thank you for claryfing this but from this `the subscription is cancelled regardless of the subscribers` i thought it refers to the current instance , the target `ISubscriber` not the entire `ISubscriber`  group of objects that listen to the target channel.

**Comment by ExploreNcrack** [comment]
> To quote from the intellisense:
> 
> ```
>      /// <summary>
>     /// Unsubscribe from a specified message channel; note; if no handler is specified, the subscription is cancelled regardless
>     /// of the subscribers; if a handler is specified, the subscription is only cancelled if this handler is the 
>     /// last handler remaining against the channel
>     /// </summary>
>     /// <param name="channel">The channel that was subscribed to.</param>
>     /// <param name="handler">The handler to no longer invoke when a message is received on <paramref name="channel"/>.</param>
>     /// <param name="flags">The command flags to use.</param>
>     /// <remarks>https://redis.io/commands/unsubscribe</remarks>
>     /// <remarks>https://redis.io/commands/punsubscribe</remarks>
>     void Unsubscribe(RedisChannel channel, Action<RedisChannel, RedisValue> handler = null, CommandFlags flags = CommandFlags.None);
> ```
> 
> Basically, to get what you want (tracked unsubscription), you should supply the same handler to `Unsubscribe` as you did to `Subscribe`; when the system determines that all handlers have been removed: it will nuke the subscription. By calling `Unsubscribe[Async]` without a handler, you're issuing a blanket unsubscribe.

Hi, can you give me an example of how to create such handler that can be referenced by both `Subscribe()` and `Unsubscribe()`?
Thank you

**Comment by NickCraver** [comment]
Closing this out to tidy up - thanks for the updates!
